Userland/Libraries/LibWeb/Bindings/IDLAbstractOperations.h

```diff
@@ -9,10 +9,175 @@
+    relevant_settings.clean_up_after_running_script();
+
+    // 3. If completion is a normal completion, return completion.
+    if (completion.type() == JS::Completion::Type::Normal)
+        return completion;
+
+    // 4. If completion is an abrupt completion and the operation has a return type that is not a promise type, return completion.
+    // FIXME: This does not handle promises and thus always returns completion at this point.
+    return completion;
+
+    // FIXME: 5. Let rejectedPromise be ! Call(%Promise.reject%, %Promise%, «completion.[[Value]]»).
+
+    // FIXME: 6. Return the result of converting rejectedPromise to the operation’s return type.
+}
+
+// https://webidl.spec.whatwg.org/#call-a-user-objects-operation
+template<typename... Args>
+JS::Completion call_user_object_operation(Bindings::CallbackType& callback, String const& operation_name, Optional<JS::Value> this_argument, Args&&... args)
+{
+    // 1. Let completion be an uninitialized variable.
+    JS::Completion completion;
+
+    // 2. If thisArg was not given, let thisArg be undefined.
+    if (!this_argument.has_value())
+        this_argument = JS::js_undefined();
+
+    // 3. Let O be the ECMAScript object corresponding to value.
+    auto& object = *callback.callback.cell();
+
+    // 4. Let realm be O’s associated Realm.
+    auto& global_object = object.global_object();
+    auto& realm = *global_object.associated_realm();
+
+    // 5. Let relevant settings be realm’s settings object.
+    auto& relevant_settings = verify_cast<HTML::EnvironmentSettingsObject>(*realm.host_defined());
+
+    // 6. Let stored settings be value’s callback context.
+    auto& stored_settings = callback.callback_context;
+
+    // 7. Prepare to run script with relevant settings.
+    relevant_settings.prepare_to_run_script();
+
+    // 8. Prepare to run a callback with stored settings.
+    stored_settings.prepare_to_run_callback();
+
+    // 9. Let X be O.
+    auto* actual_function_object = &object;
+
+    // 10. If ! IsCallable(O) is false, then:
+    if (!object.is_function()) {
+        // 1. Let getResult be Get(O, opName).
+        auto get_result = object.get(operation_name);
+
+        // 2. If getResult is an abrupt completion, set completion to getResult and jump to the step labeled return.
+        if (get_result.is_throw_completion()) {
+            completion = get_result.throw_completion();
+            return clean_up_on_return(stored_settings, relevant_settings, completion);
+        }
+
+        // 4. If ! IsCallable(X) is false, then set completion to a new Completion{[[Type]]: throw, [[Value]]: a newly created TypeError object, [[Target]]: empty}, and jump to the step labeled return.
+        if (!get_result.value().is_function()) {
+            completion = JS::Completion { JS::Completion::Type::Throw, JS::TypeError::create(global_object, String::formatted(JS::ErrorType::NotAFunction.message(), get_result.value().to_string_without_side_effects())), {} };
```

Comment:
https://github.com/SerenityOS/serenity/pull/10478#discussion_r732274377

> I'm not sure this works as intended - currently you still have to actually throw an exception in LibJS, a throw completion on its own won't create an exception (yet).
>
> `VM::throw_completion()` can do this for you:

```suggestion
            completion = realm.vm().template throw_completion<TypeError>(global_object, JS::ErrorType::NotAFunction, get_result.value().to_string_without_side_effects());
```